posts/2020-03-10-calculating-pi-concurrently.adoc

```diff
@@ -0,0 +1,926 @@
+
+All source code shown in this article can be found https://github.com/eloots/calculating-pi-concurrently[here].
+
+=== Introduction
+With the prevalence of multi-CPU computers and multi-core CPUs, the opportunity
+exists to speed-up the execution of applications in general and algorithms that
+lend themselves to parallelisation specifically.
+
+If we look at the latter, we may find that the algorithm can be decomposed into
+subtasks which can be executed in parallel. Getting the final result then
+consists of combining the sub-results of the subtasks. There we can distinguish
+two cases. In the first case, the order in which the sub-results are composed
+doesn't matter, whereas in the second case, it does.
+
+An example of the first case is when the sub-results are numbers and the final
+result is their sum. Obviously, the order in which the numbers are added
+doesn't change the result (to be complete, effects of rounding due to finite
+precision _can_ impact the final result, but that's an interesting but entirely
+different topic). An example where order definitely matters is if we multiply
+a series of matrices: if we split the multiplication into groups, calculate
+the product of the matrices in each group, then the order in which we combine
+the sub-results _does_ matter!
+
+In Scala, we have different tools available to write applications that execute
+code concurrently, and if we execute this code on a multi-CPU or multi-core
+computer, we can achieve parallelism in the execution of this code.
+
+A first tool we have is Akka: Akka allows us to create so-called Actors that
+can execute code. This execution can be triggered by sending an Actor a message.
+In the context of concurrent execution of code, we can imagine a system where
+we create a pool of Actors that execute code concurrently.
+This execution model allows us to implement concurrent execution of sub-tasks
+that fall in the first category mentioned earlier. Unless we take specific
+measures, we have no control over the order in which sub-results are returned.
+Another potential issue with an Actor based implementation is that we don't have
+any flow control. What this means is that if we throw more work at an Actor
+than this Actor can process, work will get piled-up in a buffer (actually,
+the Actor's mailbox). The conclusion is that the Actor model can be applied in a
+straightforward way to problems of the first type, albeit that we need to be
+aware of potential ordering and flow control issues. We'll come back to this
+topic later when we discuss solutions to our sample problem based on
+https://doc.akka.io/docs/akka/current/stream[Akka Streams].
+
+A second tool we have at our disposal is Scala's `Future`. This API allows us
+to tackle both of the aforementioned cases.
+
+Let's see how `Future` can be used to implement a parallelised computation of
+the number π.
+
+== Calculating π
+
+Ok, this topic certainly has been beaten to death, but it's easy to understand,
```

Comment:
I for one have never before seen anything about how pi is calculated, somehow. (Other than approximations.) Is it worth expanding on this a bit more?

(On the other hand, what I have seen beaten to death is computing factorials and fibonacci numbers. But those are not as parallelizable I suppose?)

Reply:
Interesting... Calculating π has a long history starting with the ancient Greeks. See [this graph](https://en.wikipedia.org/wiki/File:Record_pi_approximations.svg) for the historical 'growth' of calculated digits over time. [The article on the number π](https://en.wikipedia.org/wiki/Pi#Fourier_series) makes for some interesting reading.
Regarding factorials and fibonacci numbers. For factorial, there's [Stirling's approximation](https://en.wikipedia.org/wiki/Stirling%27s_approximation) and for Fibonacci, there's actually [a closed form formula](https://en.wikipedia.org/wiki/Jacques_Philippe_Marie_Binet#Binet's_Fibonacci_number_formula)!